### PR TITLE
[web][api][SIMS #1229] Ministry Views Scholastic Standing UI changes#1229

### PR DESCRIPTION
**Pull request by ann-aot**
- created a common transformation method `transformToActiveApplicationDataAPIOutDTO` for `ActiveApplicationDataAPIOutDTO` to avoid duplication.
- created a route controller, web HTTP client, and services for scholastic standing and moved existing methods, and updated the endpoints.
- unsuccessful scholastic standing does not have assessments, so the row is taken from the scholastic standing table and added to the top of the history data table.
- unsuccessful completions will not have an assessment button in the history table and all other scholastic standings will have an assessment button.
- For the readonly scholastic standing form, "ActiveApplicationDataAPIOutDTO" used in "report a change" form on the time of submission and scholastic standing submission dto "ScholasticStandingSubmittedDetailsAPIOutDTO" is merged and created a new dto called  "ScholasticStandingSubmittedAPIOutDTO"

**Screenshots**
- image
image
image
image
**swagger**,
image

### CODE REVIEW

File: sources/packages/api/src/route-controllers/institution-locations/models/application.dto.ts [hunk 1]
```diff
@@ -30,3 +40,43 @@ export class ActiveApplicationSummaryAPIOutDTO {
...
+ */
+export const transformToActiveApplicationDataAPIOutDTO = (
+  application: Application,
+  offering: EducationProgramOffering,
+): ActiveApplicationDataAPIOutDTO => {
+  return {
+    applicationStatus: application.applicationStatus,
+    applicationNumber: application.applicationNumber,
+    applicationOfferingIntensity: offering.offeringIntensity,
+    applicationOfferingStartDate: dateString(offering.studyStartDate),
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/institution-locations/models/application.dto.ts:58]: Not really part of this PR but we need to change the date format used by `dateString` to the same from Vue that we recently updated to be exactly what is requested by Figma.

File: sources/packages/api/src/route-controllers/student-scholastic-standings/models/student-scholastic-standings.dto.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { IntersectionType } from "@nestjs/swagger";
+import { IsNotEmptyObject } from "class-validator";
+import { ActiveApplicationDataAPIOutDTO } from "../../../route-controllers/institution-locations/models/application.dto";
+
+export class ScholasticStandingDataAPIInDTO {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/student-scholastic-standings/models/student-scholastic-standings.dto.ts:5]: As we talked about, let's convert it to an interface with necessary API properties only to allow the form to be changed dynamically.

File: sources/packages/api/src/route-controllers/student-scholastic-standings/student-scholastic-standings.aest.controller.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import { Controller, Get, Param, ParseIntPipe } from "@nestjs/common";
+import { ApiNotFoundResponse, ApiTags } from "@nestjs/swagger";
+import { UserGroups } from "../../auth/user-groups.enum";
+import { AuthorizedParties } from "../../auth/authorized-parties.enum";
+import { AllowAuthorizedParty, Groups } from "../../auth/decorators";
+import BaseController from "../BaseController";
+import { ClientTypeBaseRoute } from "../../types";
+import { ScholasticStandingSubmittedDetailsAPIOutDTO } from "./models/student-scholastic-standings.dto";
+
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/student-scholastic-standings/student-scholastic-standings.aest.controller.ts:9]: Please remove the extra line.

File: sources/packages/api/src/route-controllers/student-scholastic-standings/student-scholastic-standings.aest.controller.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+   * Get Scholastic Standing submitted details.
+   * @param scholasticStandingId scholastic standing id.
+   * @returns Scholastic Standing.
+   */
+  @Get(":scholasticStandingId")
+  @ApiNotFoundResponse({
+    description: "Scholastic Standing not found.",
+  })
+  async getScholasticStanding(
+    @Param("scholasticStandingId", ParseIntPipe) scholasticStandingId: number,
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/student-scholastic-standings/student-scholastic-standings.aest.controller.ts:36]: ParseIntPipe ❤️

File: sources/packages/api/src/route-controllers/student-scholastic-standings/student-scholastic-standings.institutions.controller.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      // Start assessment.
+      if (scholasticStanding.studentAssessment) {
+        await this.studentAssessmentService.startAssessment(
+          scholasticStanding.studentAssessment.id,
+        );
+      }
+    } catch (error: unknown) {
+      if (error instanceof CustomNamedError) {
+        switch (error.name) {
+          case APPLICATION_NOT_FOUND:
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/student-scholastic-standings/student-scholastic-standings.institutions.controller.ts:99]: Because `:applicationId` is part of the URI, should `APPLICATION_NOT_FOUND` be a not found exception?

File: sources/packages/web/src/services/ScholasticStandingService.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import ApiClient from "@/services/http/ApiClient";
+import {
+  ScholasticStandingDataAPIInDTO,
+  ScholasticStandingSubmittedDetailsAPIOutDTO,
+} from "./http/dto";
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/ScholasticStandingService.ts:5]: Please change to "@".

File: sources/packages/web/src/services/ScholasticStandingService.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    scholasticStandingId: number,
+  ): Promise<ScholasticStandingSubmittedDetailsAPIOutDTO> {
+    return ApiClient.ScholasticStandingApi.getScholasticStanding(
+      scholasticStandingId,
+    );
+  }
+
+  /**
+   * Save scholastic standing and create new assessment.
+   * @param applicationId application id
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/ScholasticStandingService.ts:33]: Please add periods for all sentences.

File: sources/packages/web/src/services/http/ScholasticStandingApi.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
+import HttpBaseClient from "@/services/http/common/HttpBaseClient";
+import {
+  ScholasticStandingDataAPIInDTO,
+  ScholasticStandingSubmittedDetailsAPIOutDTO,
+} from "./dto";
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/http/ScholasticStandingApi.ts:5]: Please add the "@".

File: sources/packages/web/src/services/http/ScholasticStandingApi.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    scholasticStandingId: number,
+  ): Promise<ScholasticStandingSubmittedDetailsAPIOutDTO> {
+    return this.getCallTyped<ScholasticStandingSubmittedDetailsAPIOutDTO>(
+      this.addClientRoot(`scholastic-standing/${scholasticStandingId}`),
+    );
+  }
+
+  /**
+   * Save scholastic standing and create new assessment.
+   * @param applicationId application id
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/http/ScholasticStandingApi.ts:26]: Please add periods for all sentences.

File: sources/packages/web/src/views/aest/student/ViewScholasticStanding.vue [hunk 10]
```diff
@@ -0,0 +1,60 @@
+<template>
+  <header-navigator
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/student/ViewScholasticStanding.vue:2]: The `full-page-container` has a specific slot for the header (#header), please move it inside.

File: sources/packages/web/src/views/aest/student/ViewScholasticStanding.vue [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    title="Assessments"
+    :routeLocation="goBackRouteParams"
+    subTitle="View Submission"
+  />
+  <full-page-container class="p-m-4">
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/student/ViewScholasticStanding.vue:7]: Please remove the `class="p-m-4"` it is not supposed to be needed.

**Review by andrewsignori-aot — COMMENTED** [review]
Amazing work @ann-aot and nice API organization. Please take a look at the minor comments.

File: sources/packages/api/src/route-controllers/student-scholastic-standings/student-scholastic-standings.institutions.controller.ts (hunk 5 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/student-scholastic-standings/student-scholastic-standings.institutions.controller.ts:99]: Agree 👍  not part of this PR. but i will adjust it

Comment (ann-aot) [reply sources/packages/api/src/route-controllers/student-scholastic-standings/student-scholastic-standings.institutions.controller.ts:99]: not part of this PR.  This error is throwing from the service, 
image, but the code `APPLICATION_NOT_FOUND` doesn't seem to be logical. i will update the code

File: sources/packages/api/src/route-controllers/assessment/assessment.aest.controller.ts [hunk 12]
```diff
@@ -87,23 +93,40 @@ export class AssessmentAESTController extends BaseController {
    */
   @Get("application/:applicationId/history")
   async getAssessmentHistorySummary(
```
Comment (henryoforeh-aot) [inline sources/packages/api/src/route-controllers/assessment/assessment.aest.controller.ts:95]: Is this method not supposed to support pagination ?

File: sources/packages/api/src/route-controllers/assessment/models/assessment.dto.ts [hunk 13]
```diff
@@ -21,14 +21,16 @@ export class RequestAssessmentSummaryAPIOutDTO {
...
   submittedDate: Date;
   triggerType: AssessmentTriggerType;
-  assessmentDate: Date;
+  assessmentDate?: Date;
   status: StudentAssessmentStatus;
   studentAppealId?: number;
   applicationExceptionId?: number;
   studentScholasticStandingId?: number;
+  // This flag decides, the row is unsuccessful week or not.
+  unsuccessfulWeek?: boolean;
```
Comment (henryoforeh-aot) [inline sources/packages/api/src/route-controllers/assessment/models/assessment.dto.ts:33]: I would rename to hasUnsuccessfulWeeks.

File: sources/packages/api/src/route-controllers/institution-locations/models/application.dto.ts [hunk 14]
```diff
@@ -30,3 +40,43 @@ export class ActiveApplicationSummaryAPIOutDTO {
   applicationStatus: ApplicationStatus;
   fullName: string;
 }
+
+/**
+ * Transformation util for Active application.
+ * @param application application object.
+ * @param offering offering object.
+ * @returns active application data dto
```
Comment (guru-aot) [inline sources/packages/api/src/route-controllers/institution-locations/models/application.dto.ts:48]: Please add a period on the return.

**Review by guru-aot — APPROVED** [review]
Minor comment , Nice work @ann-aot

**Review by dheepak-aot — APPROVED** [review]
Great work @ann-aot 👍. Have 2 comments which already @andrewsignori-aot mentioned on Interface (ScholasticStandingDataAPIInDTO) and using the full-page-container as root-layout.

File: sources/packages/api/src/route-controllers/assessment/assessment.aest.controller.ts (hunk 12 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/assessment/assessment.aest.controller.ts:95]: no @henryoforeh-aot in a real scenario, the history and requested summary will have only a few rows, so there is no need for lazy loading

File: sources/packages/api/src/route-controllers/assessment/models/assessment.dto.ts (hunk 13 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/assessment/models/assessment.dto.ts:33]: i will add `isUnsuccessfulWeek`

File: sources/packages/api/src/route-controllers/assessment/assessment.aest.controller.ts (hunk 12 above)
Comment (henryoforeh-aot) [reply sources/packages/api/src/route-controllers/assessment/assessment.aest.controller.ts:95]: Okay thanks for providing some more info on it. I will resolve.

**Review by henryoforeh-aot — APPROVED** [review]

File: sources/packages/api/src/route-controllers/assessment/models/assessment.dto.ts (hunk 13 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/assessment/models/assessment.dto.ts]: I would agree with `hasUnsuccessfulWeeks`.

**Review by andrewsignori-aot — COMMENTED** [review]
Please take a look at the comments.

**Review by andrewsignori-aot — APPROVED** [review]
Great work and thanks so much for doing the changes, Looks good 👍
